server.py

```diff
@@ -15,7 +15,7 @@
 # Load the configuration.
 META = json.load(open(os.path.join(DIR, 'config.json')))
 
-# Our nonce storage: a dict indexed by email. Normally, you'd store nonces with
+# Our nonce storage: a dict indexed by IP. Normally, you'd store nonces with
```

Comment:
I think this is the wrong direction. I think it's a strong invariant for the broker to always return the same email address as requested by the RP, anyway; if normalization is needed, this should be handled in the broker, hiding it from the RP.

Since normalization is probably not even consistent across hosts, I think this is a requirement for the broker, so it makes sense to rely on that.